### ISSUE TITLE
Field label is sometimes displayed even when field is empty (for some fields including: Text)

### ISSUE THREAD

**Issue by Radcliffe** [backdrop/backdrop-issues#4152]
**Description of the bug**
The label is displayed for a text field, even though the field is (or should be) empty.

**Steps To Reproduce**
1. Add a custom text field to Post or another content type.
2. Modify the default display mode to show the label for the custom field (above or inline).
3. Create a new Post.
4. View the new Post.

**Actual behavior**
The label on the empty field is displayed when viewing the new Post, but not when viewing an old Post. 

![Your_Second_Post___Demo](https://user-images.githubusercontent.com/676260/67302516-9c004780-f4b6-11e9-8aaa-204cf65c82fd.png)

The label is not displayed on an old Post that was created before the custom field was added. If I close comments on the new Post then the label is no longer displayed, and the label remains hidden when comments are reopened.

**Expected behavior**
The label should not be displayed if the field is empty.

**Additional information**
Backdrop CMS version: 1.14.1, no custom themes or modules.

**Comment by stpaultim** [comment]
I just followed the steps to reproduce using a Tugboat demo site and was able to confirm this behavior. 

This seems like it would be a pretty commonly experience problem.

**Comment by stpaultim** [comment]
In further testing, I'm noticing a few things. 

1) This only seems to happen with certain types of fields. I have confirmed it does happen with text fields, and link fields. It does not happen for me with a date field. 
2) If you resave the node, then the labels for empty fields no longer appear. So they appear, only after the first save of the node. (Does that suggest something with caching?)

**Comment by Radcliffe** [comment]
I used the Devel tab to examine the node. Initially, the field is represented by the following array, and the field label is shown.

```php
array(
  'und' => array(
     '0' => array(
       'value' => '',
       'format' => NULL,
       'safe_value' => ''
      )
  )
);
```
After resaving the node, the field is represented by an empty array, and the field label is hidden.

**Comment by oadaeh** [comment]
So how would I work around this "fix" if I wanted the label displayed, regardless of whether there is content in the field or not?

**Comment by Radcliffe** [comment]
@oadaeh: Create a text field with a single space as the default value.

**Comment by stpaultim** [comment]
@oadaeh - I am quite sure that the expected behavior is that neither the field or label should not show up if there is no content in the field. In fact, right now that is what happens after you resave a node for the text fields and for other fields it works that way all the time. 

If you want empty fields to show up reliably, I think you'll need to put in a feature request. :-) Maybe add an option to the field label display configuration?

I agree that there is a use case for this. I suspect that there are work-a-rounds. The easiest option would be provide a default value for the field and require it. 

I found this Drupal 7 module that seems to address that problem:
https://www.drupal.org/project/empty_fields

I'm sure that there are other options.

**Comment by jenlampton** [comment]
I think [my new issue](https://github.com/backdrop/backdrop-issues/issues/4473#issuecomment-656318798) is a duplicate of this one, but I ran into the same thing recently. 

@Radcliffe  what modules are you using on your site?

Here's my list:
* addressfield
* aggregator
* auto_nodetitle
* backup_migrate
* biblio
* colorbox
* devel
* extlink
* feeds
* feeds_tamper
* feeds_xpathparser
* field_group
* filefield_sources
* flexslider
* geocoder
* geofield
* geophp
* googleanalytics
* iframe
* image_resize_filter
* imce
* job_scheduler
* menu_position
* multifield
* on_the_web
* rabbit_hole
* reference
* site_verify
* special_menu_items
* textformatter
* xmlsitemap
* youtube

**Comment by klonos** [comment]
As mentioned in https://www.drupal.org/project/drupal/issues/1283974, in the `content-field.tpl.php` file in CCK 6.x, everything was wrapped in this check:

```php
<?php if (!$field_empty) : ?>
```

(see https://git.drupalcode.org/project/cck/-/blob/6.x-3.x/theme/content-field.tpl.php#L27)

In both D7 and Backdrop, this check has been removed from the respective `field.tpl.php` file though. We only check for `<?php if (!$label_hidden): ?>`.

D7: https://git.drupalcode.org/project/drupal/-/blob/7.x/modules/field/theme/field.tpl.php#L56
Backdrop: https://github.com/backdrop/backdrop/blob/1.x/core/modules/field/templates/field.tpl.php#L54

I tried to have a go at this, but I cannot reproduce the issue to begin with 🤷 ...I'm starting to believe that this must be due to some contrib.

**Comment by jenlampton** [comment]
Drupal 7 and Backdrop have render arrays. If there's not a value in the
render array, it shouldn't render anything (so what's in a field template
shouldn't matter -- it shouldn't even be used). We could add a check in
there -- or in theme_field (which is what is used until overridden with a
template, in a theme) as in my example, but I'd rather we stop the render
process earlier on and save some processing power.

I also wasn't able to reproduce this on a stock install w/o contrib, but I
want to do some deeper testing on my problem site to see If I can track
down the root cause.

**Comment by indigoxela** [comment]
I'd like to share a suspicion here. To verify, check the json export of that field instance. Copy it somewhere for later comparison.

Then save the field once via UI (without any changes) and check the json export again. Is it different now? Especially regarding `"default_value"`.

Background to my suspicion: when fixing one of the field label bugs, I realized that it might be necessary to create an update hook, but we finally didn't add it.

It might be that the fields now showing up labels even when empty have been created or at least saved once between the two changes - the on that broke usage of field_is_empty (#3378) and the one that fixed it (#3873).

At least that would explain why only *some* fields are affected while others aren't.

**Comment by indigoxela** [comment]
I'd like to share a suspicion here. To verify, check the json export of that field instance. Copy it somewhere for later comparison.

Then save the field once via UI (without any changes) and check the json export again. Is it different now? Especially regarding `"default_value"`.

Background to my suspicion: when fixing one of the field label bugs, I realized that it might be necessary to create an update hook, but we finally didn't add it.

It might be that the fields now showing up labels even when empty have been created or at least saved once between the two changes - the on that broke usage of field_is_empty (#3378) and the one that fixed it (#3873).

At least that would explain why only *some* fields are affected while others aren't.

**Comment by jenlampton** [comment]
I think you're exactly right @indigoxela I checked the config and some fields had a default value of `NULL` and others had a default value of `""` (empty string). This site got a lot of attention between April and December last year.

**Comment by stpaultim** [comment]
Well, I noted above (Oct 2019) that I was able to recreate this problem with the provided steps on a fresh Tugboat site. 

As of today, I am no longer able to reproduce this problem on a fresh site. I checked because @jenlampton was unable to reproduce this problem on fresh site.

**Comment by klonos** [comment]
Seems that we should close this then. Right?

**Comment by stpaultim** [comment]
> Seems that we should close this then. Right?

I would hold off on closing this issue, since it appears that @jenlampton is experiencing this problem on a existing site. We just don't have steps to reproduce it reliably anymore.

**Comment by jenlampton** [comment]
I have not been able to reproduce this. I think the site I was working on had its fields changed between the two issues...

> the one that broke usage of field_is_empty (#3378) and the one that fixed it (#3873).

And now that I have changed them all back, everything is fine. I think it's safe to close this issue now.
